parsers/CA_SK.py

```diff
@@ -1,142 +1,108 @@
+    session = session or Session()
+    # Mimic a user browser in the headers or the API will respond with a 403.
+    response = session.get(url, headers={"user-agent": USER_AGENT})
+    if not response.ok:
         raise ParserException(
             "CA_SK.py",
-            "This parser is unable to fetch historical data.",
+            f"Request to {url} failed. Response Code: {response.status_code}\n"
+            f"Error:\n{response.text}",
             zone_key,
         )
+    return response.json()
 
 
 def fetch_production(
-    zone_key: str = "CA-SK",
+    zone_key: ZoneKey = ZoneKey("CA-SK"),
     session: Session | None = None,
     target_datetime: datetime | None = None,
     logger: Logger = getLogger(__name__),
-):
-    """This parser function will currently return the daily average of the day in question as hourly data.
-    This is because the API only returns daily data but the backend expects hourly values.
-    This is in order to facilitate the estimation of the hourly values from the daily average.
-    """
-    # Validate that the zone key is equal to CA-SK.
-    validate_zone_key(zone_key)
-    # Validate that the target_datetime is None as this parser is unable to fetch historical data.
-    validate_no_datetime(target_datetime, zone_key)
-
-    session = session or Session()
-
-    # Set the headers to mimic a user browser as the API will return a 403 if not.
-    headers = {"user-agent": USER_AGENT}
-    response: Response = session.get(PRODUCTION_URL, headers=headers)
-
-    if not response.ok:
-        raise ParserException(
-            "CA_SK.py",
-            f"Failed to fetch production data. Response Code: {response.status_code}\nError:\n{response.text}",
-            zone_key,
-        )
-
-    raw_data = response.json()
+) -> list[dict[str, typing.Any]]:
+    payload = _request(session, target_datetime, PRODUCTION_URL, zone_key)
     # Date is in the format "Jan 01, 2020"
-    raw_date = raw_data["SupplyDataText"]
-    date = datetime.strptime(raw_date, "%b %d, %Y")
-    production_data = {}
-
-    for value in raw_data["PowerCacheData"]["generationByType"]:
-        production_data[PRODUCTION_MAPPING[value["type"]]] = value[
-            "totalGenerationForType"
-        ]
-
-    data_list: list[dict] = []
-    # Hack to return hourly data from daily data for the backend as it expects hourly data.
-    for hour in range(0, 24):
-        data_list.append(
-            {
-                "zoneKey": zone_key,
-                "datetime": date.replace(hour=hour, tzinfo=TIMEZONE),
-                "production": production_data,
-                "source": "saskpower.com",
-            }
+    date = datetime.strptime(payload["SupplyDataText"], "%b %d, %Y")
+    production_mix = ProductionMix(
+        **{
+            PRODUCTION_MAPPING[value["type"]]: value["totalGenerationForType"]
+            for value in payload["PowerCacheData"]["generationByType"]
+        }
+    )
```

Comment:
This is great and simple but due to how the ProductionMix class works it would be better to use the add_value() function of the class and use that inside the loop.

This makes sure all the data sanitation and validation runs on each value and makes it more in line with how other parsers are working / will work.